src/core/sensor/qgsiodevicesensor.cpp

```diff
@@ -0,0 +1,328 @@
+/***************************************************************************
+                             qgsiodevicesensor.cpp
+                             ---------------------------
+    begin                : March 2023
+    copyright            : (C) 2023 by Mathieu Pellerin
+    email                : mathieu at opengis dot ch
+ ***************************************************************************/
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsiodevicesensor.h"
+#include "qgssensorregistry.h"
+
+#include <QDomElement>
+
+#if defined( Q_OS_ANDROID ) || defined( Q_OS_LINUX )
+#include <sys/socket.h>
+#endif
+
+QgsIODeviceSensor::~QgsIODeviceSensor()
+{
+  delete mIODevice;
+}
+
+void QgsIODeviceSensor::initiateIODevice( QIODevice *device )
+{
+  mIODevice = device;
+  connect( mIODevice, &QIODevice::readyRead, this, &QgsIODeviceSensor::parseData );
+}
+
+void QgsIODeviceSensor::parseData()
+{
+  mData.lastValue = mIODevice->readAll();
```

Comment:
if ( !mIODevice )
  return

Reply:
@nyalldawson , is it needed here? This is a protected slot function that's ever only called when mIODevice emits the readyRead signal.